File: DMP'25/Stark Invoice/src/services/serverFiles.ts
const API_BASE_URL =
  import.meta.env.VITE_API_BASE_URL || "http://localhost:8888"; // Use environment variable with fallback

export interface ServerFile {
  id: number;
  filename: string;
  s3_key: string;
  created_at: string;
  file_size: number;
}

export interface AuthResponse {
  success: boolean;
  token: string;
  user: {
    id: number;
    name: string;
    email: string;
  };
}

export interface LoginCredentials {
  email: string;
  password: string;
}

export interface RegisterCredentials {
  name: string;
  email: string;
  password: string;
}

class ServerFilesService {
  private token: string | null = null;

  setToken(token: string) {
    this.token = token;
    localStorage.setItem("server_auth_token", token);
  }

  getToken(): string | null {
    if (!this.token) {
      this.token = localStorage.getItem("server_auth_token");
    }
    return this.token;
  }

  clearToken() {
    this.token = null;
    localStorage.removeItem("server_auth_token");
  }

  private getHeaders(): HeadersInit {
    const headers: HeadersInit = {
      "Content-Type": "application/json",
    };

    const token = this.getToken();
    if (token) {
      headers["Authorization"] = `Bearer ${token}`;
    }

    return headers;
  }

  async login(credentials: LoginCredentials): Promise<AuthResponse> {
    const response = await fetch(`${API_BASE_URL}/auth/login`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(credentials),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Login failed");
    }

    const data = await response.json();
    this.setToken(data.token);
    return data;
  }

  async register(
    credentials: RegisterCredentials
  ): Promise<{ success: boolean; message: string; user_id: number }> {
    const response = await fetch(`${API_BASE_URL}/auth/register`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(credentials),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Registration failed");
    }

    return await response.json();
  }

  async getFiles(): Promise<ServerFile[]> {
    const response = await fetch(`${API_BASE_URL}/server-files`, {
      method: "GET",
      headers: this.getHeaders(),
    });

    if (!response.ok) {
      if (response.status === 401) {
        this.clearToken();
        throw new Error("Authentication required");
      }
      const error = await response.json();
      throw new Error(error.error || "Failed to fetch files");
    }

    const data = await response.json();
    return data.files;
  }

  async uploadFile(file: File): Promise<{
    success: boolean;
    message: string;
    file_id: number;
    filename: string;
  }> {
    const formData = new FormData();
    formData.append("file", file);

    const headers: HeadersInit = {};
    const token = this.getToken();
    if (token) {
      headers["Authorization"] = `Bearer ${token}`;
    }

    const response = await fetch(`${API_BASE_URL}/server-files/upload`, {
      method: "POST",
      headers,
      body: formData,
    });

    if (!response.ok) {
      if (response.status === 401) {
        this.clearToken();
        throw new Error("Authentication required");
      }
      const error = await response.json();
      throw new Error(error.error || "Upload failed");
    }

    return await response.json();
  }

  async uploadInvoiceData(
    filename: string,
    content: string,
    billType: number
  ): Promise<{
    success: boolean;
    message: string;
    file_id: number;
    filename: string;
  }> {
    // Create a JSON file with the invoice data
    const fileData = {
      fileName: `server_${filename}`,
      content: content,
      timestamp: new Date().toISOString(),
      billType: billType,
    };

    // Create a Blob from the JSON data
    const jsonBlob = new Blob([JSON.stringify(fileData, null, 2)], {
      type: "application/json",
    });

    // Create a File object from the Blob
    const file = new File([jsonBlob], `server_${filename}.json`, {
      type: "application/json",
    });

    // Use the existing uploadFile method
    return await this.uploadFile(file);
  }

  async downloadFile(fileId: number): Promise<Blob> {
    const response = await fetch(
      `${API_BASE_URL}/server-files/download/${fileId}`,
      {
        method: "GET",
        headers: this.getHeaders(),
      }
    );

    if (!response.ok) {
      if (response.status === 401) {
        this.clearToken();
        throw new Error("Authentication required");
      }
      if (response.status === 404) {
        throw new Error("File not found");
      }
      const error = await response.json();
      throw new Error(error.error || "Download failed");
    }

    return await response.blob();
  }

  async deleteFile(
    fileId: number
  ): Promise<{ success: boolean; message: string }> {
    const response = await fetch(
      `${API_BASE_URL}/server-files/delete/${fileId}`,
      {
        method: "DELETE",
        headers: this.getHeaders(),
      }
    );

    if (!response.ok) {
      if (response.status === 401) {
        this.clearToken();
        throw new Error("Authentication required");
      }
      if (response.status === 404) {
        throw new Error("File not found");
      }
      const error = await response.json();
      throw new Error(error.error || "Delete failed");
    }

    return await response.json();
  }

  isAuthenticated(): boolean {
    return !!this.getToken();
  }
}

export const serverFilesService = new ServerFilesService();

File: DMP'25/AWS S3/apis/server_files.py
from flask import Blueprint, request, jsonify, send_file
import jwt
import uuid
import os
from services.database import cursor, conn
from services.s3 import s3_client, BUCKET_NAME
from datetime import datetime
from dotenv import load_dotenv
from io import BytesIO

# Load environment variables
load_dotenv()

# Get secret key from environment variable
SECRET_KEY = os.getenv(
    "JWT_SECRET_KEY", "your-secret-key-here-change-this-in-production")

server_files_bp = Blueprint('server_files', __name__)


def get_user_from_token(token):
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=['HS256'])
        return payload.get('user_id')
    except jwt.ExpiredSignatureError:
        return None
    except jwt.InvalidTokenError:
        return None


@server_files_bp.route('/', methods=['GET'])
def get_files():
    """Get all files for authenticated user"""
    try:
        auth_header = request.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return jsonify({"error": "Authentication required"}), 401

        token = auth_header.split(' ')[1]
        user_id = get_user_from_token(token)

        if not user_id:
            return jsonify({"error": "Invalid or expired token"}), 401

        # Get user's files from database
        cursor.execute(
            "SELECT id, filename, s3_key, created_at, file_size FROM user_files WHERE user_id = %s ORDER BY created_at DESC",
            (user_id,)
        )
        files = cursor.fetchall()

        # Convert to JSON serializable format
        files_list = []
        for file in files:
            files_list.append({
                'id': file['id'],
                'filename': file['filename'],
                's3_key': file['s3_key'],
                'created_at': file['created_at'].isoformat() if file['created_at'] else None,
                'file_size': file['file_size']
            })

        return jsonify({"files": files_list})

    except Exception as e:
        return jsonify({"error": str(e)}), 500


@server_files_bp.route('/upload', methods=['POST'])
def upload_file():
    """Upload a new file"""
    try:
        auth_header = request.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return jsonify({"error": "Authentication required"}), 401

        token = auth_header.split(' ')[1]
        user_id = get_user_from_token(token)

        if not user_id:
            return jsonify({"error": "Invalid or expired token"}), 401

        if 'file' not in request.files:
            return jsonify({"error": "No file provided"}), 400

        file = request.files['file']
        if file.filename == '':
            return jsonify({"error": "No file selected"}), 400

        file_name = file.filename
        file_content = file.read()
        content_type = file.content_type

        # Generate unique filename
        unique_filename = f"{uuid.uuid4()}-{file_name}"
        s3_key = f"user_{user_id}/{unique_filename}"

        # Upload to S3
        s3_client.put_object(
            Bucket=BUCKET_NAME,
            Key=s3_key,
            Body=file_content,
            ContentType=content_type
        )

        # Save to database
        cursor.execute(
            "INSERT INTO user_files (user_id, filename, s3_key, file_size, created_at) VALUES (%s, %s, %s, %s, %s) RETURNING id",
            (user_id, file_name, s3_key, len(file_content), datetime.utcnow())
        )
        result = cursor.fetchone()
        if result:
            file_id = result['id']
            conn.commit()
        else:
            return jsonify({"error": "Failed to save file metadata"}), 500

        return jsonify({
            "success": True,
            "message": "File uploaded successfully",
            "file_id": file_id,
            "filename": file_name
        })

    except Exception as e:
        conn.rollback()
        return jsonify({"error": str(e)}), 500


@server_files_bp.route('/download/<int:file_id>', methods=['GET'])
def download_file(file_id):
    """Download a specific file"""
    try:
        auth_header = request.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return jsonify({"error": "Authentication required"}), 401

        token = auth_header.split(' ')[1]
        user_id = get_user_from_token(token)

        if not user_id:
            return jsonify({"error": "Invalid or expired token"}), 401

        # Get file info from database
        cursor.execute(
            "SELECT filename, s3_key FROM user_files WHERE id = %s AND user_id = %s",
            (file_id, user_id)
        )
        file_info = cursor.fetchone()

        if not file_info:
            return jsonify({"error": "File not found"}), 404

        # Download from S3
        response = s3_client.get_object(
            Bucket=BUCKET_NAME, Key=file_info['s3_key'])

        # Create BytesIO object for Flask send_file
        file_data = BytesIO(response['Body'].read())
        file_data.seek(0)

        return send_file(
            file_data,
            mimetype=response['ContentType'],
            as_attachment=True,
            download_name=file_info['filename']
        )

    except Exception as e:
        return jsonify({"error": str(e)}), 500


@server_files_bp.route('/delete/<int:file_id>', methods=['DELETE'])
def delete_file(file_id):
    """Delete a specific file"""
    try:
        auth_header = request.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return jsonify({"error": "Authentication required"}), 401

        token = auth_header.split(' ')[1]
        user_id = get_user_from_token(token)

        if not user_id:
            return jsonify({"error": "Invalid or expired token"}), 401

        # Get file info from database
        cursor.execute(
            "SELECT s3_key FROM user_files WHERE id = %s AND user_id = %s",
            (file_id, user_id)
        )
        file_info = cursor.fetchone()

        if not file_info:
            return jsonify({"error": "File not found"}), 404

        # Delete from S3
        s3_client.delete_object(
            Bucket=BUCKET_NAME, Key=file_info['s3_key'])

        # Delete from database
        cursor.execute(
            "DELETE FROM user_files WHERE id = %s AND user_id = %s", (file_id, user_id))
        conn.commit()

        return jsonify({"success": True, "message": "File deleted successfully"})

    except Exception as e:
        conn.rollback()
        return jsonify({"error": str(e)}), 500
